distributed/s3.py

```diff
@@ -26,148 +32,635 @@
     if lazy:
-        values = [Value(name, [{name: (read_content_from_keys, bucket_name,
-                                       key, anon)}])
-                  for name, key in zip(names, keys)]
+        executor._send_to_scheduler({'op': 'update-graph',
+                                     'tasks': {},
+                                     'dependencies': set(),
+                                     'keys': [],
+                                     'client': executor.id})
+        values = [Value(name, [{name: (read_block_from_s3, fn, offset, length, s3pars, delimiter)}])
+                  for name, fn, offset, length in zip(names, filenames, offsets, lengths)]
         return values
     else:
-        return executor.map(read_content_from_keys, [bucket_name] * len(keys),
-                keys, anon=anon)
+        return executor.map(read_block_from_s3, filenames, offsets, lengths,
+                s3pars=s3pars, delimiter=delimiter)
 
 
-def read_text(bucket_name, prefix='', path_delimiter='', encoding='utf-8',
-        errors='strict', lineterminator='\n', executor=None, anon=None,
-        collection=True, lazy=True, compression=None):
-    """
-    Read lines of text from S3
+def read_block_from_s3(filename, offset, length, s3pars={}, delimiter=None):
+    s3 = S3FileSystem(**s3pars)
+    bytes = s3.read_block(filename, offset, length, delimiter)
+    return bytes
+
+
+def bytes_read_csv(b, **kwargs):
+    from io import BytesIO
+    import pandas as pd
+    bio = BytesIO(b)
+    return pd.read_csv(bio, **kwargs)
+
+
+def avro_body(data, header):
+    """ Convert bytes and header to Python objects
 
     Parameters
     ----------
-    bucket_name: string
-        Name of S3 bucket like ``'my-bucket'``
-    prefix: string
-        Prefix of key name to match like ``'/data/2016/``
-    path_delimiter: string (optional)
-        Delimiter like ``'/'`` to define implicit S3 directory structure
-    compression: {None, 'gzip'}
+    data: bytestring
+        bulk avro data, without header information
+    header: bytestring
+        Header information collected from ``fastavro.reader(f)._header``
 
     Returns
     -------
-    Dask bag
+    List of deserialized Python objects, probably dictionaries
     """
+    import fastavro
+    sync = header['sync']
+    if not data.endswith(sync):
+        # Read delimited should keep end-of-block delimiter
+        data = data + sync
+    stream = io.BytesIO(data)
+    schema = header['meta']['avro.schema'].decode()
+    schema = json.loads(schema)
+    codec = header['meta']['avro.codec'].decode()
+    return list(fastavro._reader._iter_avro(stream, header, codec,
+        schema, schema))
+
+
+def avro_to_df(b, av):
+    """Parse avro binary data with header av into a pandas dataframe"""
+    import pandas as pd
+    return pd.DataFrame(data=avro_body(b, av))
+
+
+@gen.coroutine
+def _read_avro(path, executor=None, fs=None, lazy=True, **kwargs):
```

Comment:
CSV I removed due to no tests (but appears to work)
Seems a pity, it works surprisingly easy. Could be kept elsewhere...